### PR TITLE
Refactor SqlQueryScheduler

### PR DESCRIPTION
**Pull request by rschlussel**
Refactorings in SqlQueryScheduler in preparation for stage retries (https://github.com/prestodb/presto/issues/13438)

* Create SqlStageExecutionInfo so we can get rid of the hacky way
  StageLinkages and StageSchedulers were created
* Separate out the parts of createStreamingLinkedStageExecutions that
  create the bucketToPartition, StageSchedulers, and child stages so
  that they aren't interleaved anymore. This refactoring changed the list of
  stageExecutions from a preorder to a postorder representation of the tree,
  so all the places that expected the root at the head of the list now look
  at the end of the list.
* Get rid of the stageExecutions, stageLinkages and
  stageSchedulers fields and instead have a single map of
  SqlStageExecutionInfos

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 1]
```diff
@@ -414,13 +397,76 @@ private static URI getBufferLocation(RemoteTask remoteTask, OutputBufferId rootB
...
+                partitioningHandle,
+                splitSources,
+                childStageExecutions);
+        StageLinkage stageLinkage = new StageLinkage(fragmentId, parent, childStageExecutions);
+        stageExecutionInfos.add(new SqlStageExecutionInfo(stageExecution, stageLinkage, stageScheduler));
+
+        return stageExecutionInfos.build();
+    }
+
+    private StageScheduler getStageScheduler(
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: nit: Should this be `createStageScheduler` ?

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 2]
```diff
@@ -1049,4 +1084,33 @@ public StreamingSubPlan withBucketToPartition(Optional<int[]> bucketToPartition)
             return new StreamingSubPlan(fragment.withBucketToPartition(bucketToPartition), children);
         }
     }
+
+    private static class SqlStageExecutionInfo
+    {
+        private final SqlStageExecution stageExecution;
+        private final StageLinkage stageLinkage;
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: Maybe add a comment stating `StageLinkage` links to the child stages? -- This is not related to this refactor but previously I need to think about whether it's up-linkage or down-linkage whenever I look into the code.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 3]
```diff
@@ -310,7 +299,7 @@ private static URI getBufferLocation(RemoteTask remoteTask, OutputBufferId rootB
         return uriBuilderFrom(location).appendPath("results").appendPath(rootBufferId.toString()).build();
     }
 
-    private List<SqlStageExecution> createStageExecutions(
+    private List<SqlStageExecutionInfo> createStageExecutions(
```
[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: The reason for a postorder representation is so we can separate out the creation of schedulers from creation of the children.  We need to have all the child stages to create the scheduler for SCALED_WRITER_DISTRIBUTION, and it's much simpler to put the subplans into the list/tree when we create them than to have to do it later.

**Comment by rschlussel** [comment]
Materialized tests keep timing out, so I think I must have introduced some deadlock. I'm looking into it.

**Comment by rschlussel** [comment]
forgot to postorder the streaming plan sections relative to each other too.  Fixing that now.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 4]
```diff
@@ -1049,4 +1084,33 @@ public StreamingSubPlan withBucketToPartition(Optional<int[]> bucketToPartition)
             return new StreamingSubPlan(fragment.withBucketToPartition(bucketToPartition), children);
         }
     }
+
+    private static class SqlStageExecutionInfo
```
[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: sure, i can't think of anything better.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java (hunk 2 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: hmmm... it links to both parent and children. it has the parent's ExchangeLocationsConsumer and the children's OutputBufferManagers.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 5]
```diff
@@ -417,13 +404,76 @@ private static URI getBufferLocation(RemoteTask remoteTask, OutputBufferId rootB
...
+            Function<PartitioningHandle, NodePartitionMap> partitioningCache,
+            NodePartitioningManager nodePartitioningManager,
+            ScheduledExecutorService schedulerExecutor,
+            Optional<SqlStageExecution> parentStageExecution,
+            StageId stageId, SqlStageExecution stageExecution,
+            PartitioningHandle partitioningHandle,
+            Map<PlanNodeId, SplitSource> splitSources,
+            Set<SqlStageExecution> childStageExecutions)
+    {
+        StageScheduler stageScheduler;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: Why is this variable needed? Can we simply return in each branch?

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 6]
```diff
@@ -582,16 +589,38 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
                     getWriterMinSize(session));
             whenAllStages(childStageExecutions, StageExecutionState::isDone)
                     .addListener(scheduler::finish, directExecutor());
-            stageSchedulers.put(stageId, scheduler);
         }
+        return stageScheduler;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: The scheduler for `SCALED_WRITER_DISTRIBUTION` is never assigned and returned.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: Let's do `if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {` as another `else if`

**Comment by rschlussel** [comment]
comments addressed and tests pass

**Pull request merged**